nemo/collections/common/prompts/formatter.py

```diff
@@ -0,0 +1,294 @@
+    and for inference (context-only).
+    Training/inference is determined automatically; if the last role in a dialog is the INFERENCE_ROLE,
+    that's an 'asked-and-answered' scenario, so we assume it's inteded for training.
+    We'll create a dict with tokenized results available under the following keys:
+
+    * ``context_ids`` (all turns minus last one),
+    * ``answer_ids`` (last turn)
+    * ``input_ids`` (previous two values concatenated)
+    * ``mask`` (boolean mask tensor of the same lenth as ``input_ids`` that's set to True on INFERENCE_ROLE turns)
+
+    Typically, the user will use the ``encode_dialog`` method providing a list of turns to it.
+    Example showing how to construct model inputs/outputs for training::
+
+        >>> formatter = PromptFormatter(tokenizer)
+        ... encoded_for_training = formatter.encode_dialog(
+        ...     turns=[
+        ...         {"role": "user", "slots": {"message": "What time is it?"}},
+        ...         {"role": "assistant", "slots": {"message": "Ten o'clock."}},
+        ...         {"role": "user", "slots": {"message": "PM or AM?"}},
+        ...         {"role": "assistant", "slots": {"message": "AM, naturally! It's bright outside"}},
+        ...     ]
+        ... )
+
+    Another example that shows how to use the same method to generate prompts for inference::
+
+
+        >>> formatter = PromptFormatter(tokenizer)
+        ... encoded_for_training = formatter.encode_dialog(
+        ...     turns=[
+        ...         {"role": "user", "slots": {"message": "What time is it?"}},
+        ...         {"role": "assistant", "slots": {"message": "Ten o'clock."}},
+        ...         {"role": "user", "slots": {"message": "PM or AM?"}},
+        ...     ]
+        ... )
+
+    """
+
+    # Used to support AggregateTokenizer; this key selects the right sub-tokenizer for each turn.
+    PROMPT_LANGUAGE_SLOT = "prompt_language"
+
+    # Sub-classes will be registered under this name, to be used via PromptFormatter.resolve(name).
+    REGISTER_NAME = None
+
+    # Template is a dict that maps:
+    # * from a role name string (system/user/assistant/etc)
+    # * to a dict with keys
+    #   * "template" that has a string value (the prompt template)
+    #   * "slots" that has a value of dict[str, Modality]
+    #       * keys of slots are the names of formattable slots in the prompt template
+    #       * values of slots are :class:`Modality` objects that can be used to check
+    #           whether a specific value conforms to a given modality requirements
+    #           (e.g., Modality.Text may expect string objects).
+    # Template is intended to be defined by the child classes.
+    TEMPLATE = None
+
+    # Turns under this role indicate responses by the model; if the last turn in
+    # PromptFormatter.encode_dialog() ends with this role, it indicates a training example.
+    INFERENCE_ROLE = None
```

Comment:
I wonder if we can use a better name than `INFERENCE_ROLE`, maybe `MODEL_ROLE`, `SYSTEM_ROLE` or something else? Personally I feel `INFERENCE_ROLE` a bit strange

Reply:
system is ambiguous because of "system prompt" in llama which is not the system's response, so I discarded that option. model - maybe? I find inference the least ambiguous. WDYT?

Reply:
hmmm.....personally I feel INFERENCE is not very straightforward, since both the model's output and user's input are taken into inference process, and I didn't understand what INFERENCE_ROLE mean until I read the documents. maybe @krishnacpuvvada @titu1994 can provide more suggestions?

Reply:
maybe ASSISTANT_ROLE or RESPONSE_ROLE...

Reply:
yeah, but these are terminologies for specific instances of LLMs such as llama. I am still hesitant about any of these names. Let me try OUTPUT_ROLE as that identifies which one is the output of whatever we call the model/system/assistant.

Reply:
Seems reasonable